waspc/test/Analyzer/ParserTest.hs

```diff
@@ -0,0 +1,157 @@
+      let source =
+            unlines
+              [ "test Imports {",
+                "  module: import Page from \"page.jsx\",",
+                "  field: import { Page } from \"page.jsx\"",
+                "}"
+              ]
+      let ast =
+            AST
+              [ Decl "test" "Imports" $
+                  Dict
+                    [ ("module", ExtImport (ExtImportModule "Page") "page.jsx"),
+                      ("field", ExtImport (ExtImportField "Page") "page.jsx")
+                    ]
+              ]
+      parse source `shouldBe` Right ast
+
+    it "Parses unary lists" $ do
+      let source = "test Decl [ 1 ]"
+      let ast = AST [Decl "test" "Decl" $ List [IntegerLiteral 1]]
+      parse source `shouldBe` Right ast
+
+    it "Parses lists of multiple elements" $ do
+      let source = "test Decl [ 1, 2, 3 ]"
+      let ast = AST [Decl "test" "Decl" $ List [IntegerLiteral 1, IntegerLiteral 2, IntegerLiteral 3]]
+      parse source `shouldBe` Right ast
+
+    it "Parses empty dictionaries and lists" $ do
+      let source = "test Decl { dict: {}, list: [] }"
+      let ast =
+            AST
+              [ Decl "test" "Decl" $
+                  Dict
+                    [ ("dict", Dict []),
+                      ("list", List [])
+                    ]
+              ]
+      parse source `shouldBe` Right ast
+
+    it "Allows trailing commas in lists and dictionaries" $ do
+      let source =
+            unlines
+              [ "test Decl {",
+                "  list: [ 1, ],",
+                "}"
+              ]
+      let ast =
+            AST
+              [ Decl "test" "Decl" $
+                  Dict
+                    [("list", List [IntegerLiteral 1])]
+              ]
+      parse source `shouldBe` Right ast
+
+    it "Parses quoted PSL" $ do
+      let source =
+            unlines
+              [ "test PSL {=psl",
+                "  id Int @id",
+                "psl=}"
+              ]
+      let ast = AST [Decl "test" "PSL" $ Quoter "psl" "\n  id Int @id\n"]
+      parse source `shouldBe` Right ast
+
+    it "Parses quoted JSON" $ do
+      let source =
+            unlines
+              [ "test JSON {=json",
+                "  \"key\": \"value\"",
+                "json=}"
+              ]
+      let ast = AST [Decl "test" "JSON" $ Quoter "json" "\n  \"key\": \"value\"\n"]
+      parse source `shouldBe` Right ast
+
+    it "Parses multiple quoters" $ do
+      let source =
+            unlines
+              [ "test JSON {=json",
+                "  \"key\": \"value\"",
+                "json=}",
+                "test JSON2 {=json",
+                "  \"key\": \"value\"",
+                "json=}"
+              ]
+      let ast =
+            AST
+              [ Decl "test" "JSON" $ Quoter "json" "\n  \"key\": \"value\"\n",
+                Decl "test" "JSON2" $ Quoter "json" "\n  \"key\": \"value\"\n"
+              ]
+      parse source `shouldBe` Right ast
+
```

Comment:
Should we add a test that parser should parse `{=foo {=bar hi bar=} foo=}` as quoter "foo" with content "{=bar hi bar=}"?
I would certainly want quoters to behave like that, although I am not sure if this would pass right now?

Reply:
That test would fail right now, but I agree that should work. Which of the following would you also expect to parse? I'll put what I expect next to each.

1. ✔️ `{=foo {=foo foo=} foo=}`
2. ❌ `{=foo foo=} foo=}`
3. ❌ `{=foo {=foo foo=}`
4. ✔️ `{=foo {=bar foo=}`
5. ✔️ `{=foo bar=} foo=}`

We should remember to modify the LaTeX document with our decisiion.

Reply:
Nice checklist :)!
I would expect following:

1. ❌ `{=foo {=foo foo=} foo=}`  -> Parses into `Quoter "foo" "{=foo foo=}"` and then fails on `foo=}` because it doesn't know how to parse that. 
2. ❌ `{=foo foo=} foo=}` -> Parses into `Quoter "foo" ""` and then fails on `foo=}` because it doesn't know how to parse that.
3. ✔️ `{=foo {=foo foo=}` -> Parses into `Quoter "foo" "{=foo"`.
4. ✔️ `{=foo {=bar foo=}` -> Parses into `Quoter "foo" "{=bar"`.
5. ✔️ `{=foo bar=} foo=}` -> Parses into `Quoter "foo" "bar=}"`.
6. ✔️ `{=foo {=bar bar=} foo=}` -> Parses into `Quoter "foo" "{=bar bar=}"`.

Assuming that we have some
```hs
type QuoterName = String
type QuoterBody = String
data Quoter = Quoter QuoterName QuoterBody
```

To put it shortly: once `{=foo` is encountered, everything is part of the body until first `foo=}` is encountered. That is it. Body is a "black box", at this stage of parsing, in the sense that we allow anything as long as we don't hit the exact `foo=}` sequence of characters.